feat : JsonValueRedisTemplate 추가 기능 구현 #12

Yundae203 opened the pull request:
> PR 당 코드의 변경은 300줄이 넘어가지 않도록 노력하기

## 📝 요약

레디스 템플릿 valueOps의 범용적 기능을 어댑터로 구현했습니다.

## 💁‍♂️ 이슈

- Json변환 실패시 예외를 일단 RuntimeError로 처리했는데, 이를 공통 커스텀 로직으로 처리할 것인지 이야기가 필요합니다.

## 🔀 변경사항

- ObjectMapper를 필드에서 생성하는 대신 생성자로 주입받기로 변경했습니다.

- 에러를 try-catch로 감싸지 않아서 외부로 유출되는 문제로 인해 메서드로 변환 부분 감싼 뒤, 에러를 처리 했습니다.

## 🔗 이슈번호

#11

Yundae203 committed:
feat: 공통 모듈 패키지 작성

각 서비스에서 공통적으로 사용할 모듈 패키지 작성 완료했습니다.

Yundae203 committed:
Merge pull request #13 from tkdwls4453/feature/#11-create-common-package

feat: 공통 모듈 패키지 작성

Yundae203 committed:
feat: JsonValueReidsTemplate 작성 #12

템플릿 공유를 위하여 중간 풀 리퀘스트 요청합니다.

Yundae203 committed:
Merge branch 'main' into feature/#12-string-redis-template

Yundae203 committed:
docs: PR_Template 파일 폴더 위치 변경 #7

Yundae203 committed:
refactor : 제너릭 수준을 클래스에서 메서드로 변경 #12

제너릭 수준을 메서드로 관리하는 것이 확장성에서 좀 더 유리할 것이라고 판단
제너릭 수준을 클래스에서 메서드로 변경
이에 따른 코드 변경사항 수정

Yundae203 committed:
Merge pull request #17 from tkdwls4453/feature/#11-create-common-package

feat : JsonValueRedisTemplate 부분 구현

Yundae203 committed:
feat : JsonValueRedisTemplate 추가 기능 구현 #12

범용적으로 쓰이는 레디스 템플릿의 기능을 어댑터로 구현 했습니다.

tkdwls4453 committed:
feat: common 모듈에 Lombok 의존성 추가

tkdwls4453 committed:
feat: API 응답 포맷 작성

API 의 일관성을 유지하기 위해서 공통 응답 포맷을 작성

참고
- https://github.com/omniti-labs/jsend
- https://github.com/tkdwls4453/ecommerce-msa-server/discussions/18

tkdwls4453 commented on common/build.gradle:
```diff
@@ -1,19 +1,34 @@
+    id 'io.spring.dependency-management' version '1.1.7'
 }
 
-group = 'com.msa'
```
여기 변경한 이유가 있을까요?

tkdwls4453 commented on common/src/main/java/com/msa/common/redis/JsonValueRedisTemplate.java:
```diff
@@ -3,25 +3,83 @@
+        return Optional.ofNullable(stringToValue(stringValue, clazz));
+    }
+
+    public void increment(String key) {
```
increament(), decrement() 메소드는 숫자 데이터 값을 증가시키고 감소시키는 것 같은데, 인자로 숫자 데이터만 들어오나요?
클래스 이름이 JsonValueRedisTemplate 이라고 되어 있어서 Json 데이터도 들어올 수 있을 거 같아서 질문드려요

younghyun-j commented on common/src/main/java/com/msa/common/redis/JsonValueRedisTemplate.java:
```diff
@@ -3,25 +3,83 @@
+        return valueOps.setIfAbsent(key, stringValue, time, timeUnit);
+    }
+
+    private <T> String valueToString(T value) {
+        try {
+            return mapper.writeValueAsString(value);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private <T> T stringToValue(String stringValue, Class<T> clazz) {
+        try {
+            return mapper.readValue(stringValue, clazz);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
     }
```
해당 코드는 Redis와 관계없이 value 매개변수를 직렬화, 역직렬화하는 코드인 것 같은데 RedisTemplate에서 제공하는 이유가 있을까요?

tkdwls4453 replied:
제 생각에는 private 접근제어자를 사용한 것으로 보아 아마 RedisTemplate 에서 제공하는 기능은 아니고 내부에서만 편하게 쓰기 위해서 만든 거 같아요

younghyun-j replied:
그렇네요! 접근제어자가 private이군요!

tkdwls4453 commented on common/src/main/java/com/msa/common/redis/JsonValueRedisTemplate.java:
```diff
@@ -3,25 +3,83 @@
+        return Optional.ofNullable(stringToValue(stringValue, clazz));
+    }
+
+    public void increment(String key) {
+        increment(key, 1);
+    }
+
+    public void increment(String key, long amount) {
+        valueOps.increment(key,amount);
+    }
+
+    public void decrement(String key) {
+        decrement(key, 1);
+    }
+
+    public void decrement(String key, long amount) {
+        valueOps.decrement(key, amount);
+    }
+
+    public boolean delete(String key) {
+        return valueOps.getOperations().delete(key);
+    }
```
increament(), decrement() 메소드는 숫자 데이터 값을 증가시키고 감소시키는 것 같은데, key에 해당하는 value가 항상 숫자 데이터인 것이 보장이 되나요? 클래스 이름이 JsonValueRedisTemplate 이라서 레디스에 Json 형태로 데이터가 있을 거 같아서 질문드립니다.

tkdwls4453 committed:
Merge pull request #21 from tkdwls4453/feature/#16-common-response

feat: 공통 응답 포맷 개발

younghyun-j commented on common/src/main/java/com/msa/common/redis/JsonValueRedisTemplate.java:
```diff
@@ -3,25 +3,83 @@
+    public <T> boolean setIfAbsent(String key, T value, Class<T> clazz, long time, TimeUnit timeUnit) {
+        String stringValue = valueToString(value);
+        return valueOps.setIfAbsent(key, stringValue, time, timeUnit);
+    }
```
매개변수 Class<T> clazz 사용 안하시면 지워도 될 것 같아요!

tkdwls4453 committed:
feat: 공통으로 사용할 상태 코드 인터페이스 정의 #16

각 서비스 마다 사용할 상태 코드의 형태를 맞추기 위해서 인터페이스를 정의했습니다.
각 서비스에서 해당 인터페이스를 Enum 형태로 구현하여 사용할 수 있습니다.

프로젝트 전반적으로 공통으로 사용될 상태 코드는 GlobalStatusCode 에 정의해 두었습니다.

tkdwls4453 committed:
refactor: ApiResponse 코드 수정 # 16

StatusCode 인터페이스를 인자로 사용하여 여러 상태 코드 구현체가 응답을 생성할 수 있도록 수정하였습니다.

tkdwls4453 committed:
feat: 프로젝트 전반적인 예외처리 #16

CustomException 을 상속받아 동일한 형태의 사용자 예외를 생성할 수 있도록 했습니다.
CustomExceptionHandler 가 사용자 정의 예외를 처리하고 그 밖의 모든 예외는 GlobalExceptionHandler 가 처리합니다.

tkdwls4453 committed:
test: 예외처리 및 에러코드 사용법 정의 #16

루트 프로젝트에 예외 처리와 에러코드를 정의하는 방법을 소개하기 위해 테스트용 코드를 작성했습니다.

코드를 확인하여 팀원들의 에러코드 및 예외처리 방식의 이해를 돕기 위해 작성하였습니다.

해당 코드는 추후에 삭제 예정

younghyun-j committed:
feat : RedisHashTemplate 구현

* RedisTemplate에서 제공하는 메서드를 추상화하여 범용적으로 사용할 수 있도록 구현

younghyun-j committed:
style : 불필요한 코드 제거

younghyun-j committed:
Merge branch 'develop' into feature/#15-hash-redis-template

tkdwls4453 committed:
Merge pull request #24 from tkdwls4453/feature/#16-global-exception

Feature/#16 global exception

younghyun-j committed:
refactor : valueToString, stringToValue 메서드 추가하여 JsonProcessingException 집중화 #15 #11

younghyun-j committed:
refactor : get 메서드 코드 간결화, increment 오버로딩 메서드 호출 변경 #15 #11

younghyun-j committed:
Merge pull request #26 from tkdwls4453/feature/#15-hash-redis-template

feat : RedisHashTemplate 구현

tkdwls4453 reviewed (approved):
👍

tkdwls4453 committed:
feat: 유저 서비스 모듈 셋팅

유저 서비스를 위한 스프링부트 프로젝트 셋팅

tkdwls4453 committed:
test: 유저 회원가입 컨트롤러 테스트 코드 작성 #25

회원 가입 요청 정보의 유효성을 체크한다.

tkdwls4453 committed:
feat: 유저 회원가입 컨트롤러 코드 작성

컨트롤러 코드만 작성, 서비스 로직은 작성하지 않음

tkdwls4453 committed:
feat: 스프링 시큐리티 의존성 추가 #25

tkdwls4453 committed:
test: 유저 회원가입 서비스 테스트 코드 작성 #25

tkdwls4453 committed:
feat: 유저 회원가입 서비스 코드 작성 #25

tkdwls4453 committed:
test: 스프링 시큐리티 추가로 기존 테스트 코드 변경 #25

tkdwls4453 committed:
style: 필요없는 import 제거 #25

tkdwls4453 committed:
chore: .gitignore 파일 수정 #25

/out 이 포함되어 /adapter/out 변경을 못 잡아서 제거함

tkdwls4453 committed:
test: adapter 유저 저장 테스트 작성

adapter 에서 유저 도메인을 받아 엔티티로 변경하여 저장하는 테스트 작성

tkdwls4453 committed:
feat: 스프링 시큐리티에서 모든 요청 통과 #25

추후에 API GATEWAY 에서 라우팅할 때 인증할 예정으로 여기서는 모두 통과시킴

tkdwls4453 committed:
feat: adapter 코드 작성

유저 도메인을 입력받아 엔티티로 변경후 데이터베이스에 저장
JPA 를 사용

tkdwls4453 committed:
chore: .gitignore 수정

tkdwls4453 committed:
chore: 도커 설정

도커 컴포즈에 유저 서비스 컨테이너와 mysql 컨테이너를 관리하는 스크립트 작성

이미지 빌드 명령어: ./gradlew docker
컨테이너 실행 명령어: docker compose up -d

tkdwls4453 committed:
test: 이메일 검증 로직 테스트 코드 추가 #25

tkdwls4453 committed:
feat: 회원가입시 이메일 중복 확인 로직 추가 #25

tkdwls4453 committed:
Merge pull request #28 from tkdwls4453/feature/#25-user-sign-up

유저 회원가입 기능 완성

tkdwls4453 committed:
chore: ci 설정 파일 작성

tkdwls4453 committed:
Merge pull request #29 from tkdwls4453/feature/ci-config

chore: ci 설정 파일 작성

Yundae203 committed:
feat: JsonValueRedisTemplate 추가 기능 구현 및 Refactor #12

불필요하게 파라미터로 받고 있던 clazz 파라미터를 제거 했습니다.
- multiGet()
- multiSet()
- size()
메서드를 추가로 구현했습니다.

Yundae203 committed:
Merge branch 'develop' into feature/#12-string-redis-template

Yundae203 commented:
수정사항 반영하겠습니다
